applications/vanilla/models/class.discussionmodel.php

```diff
@@ -2464,4 +2464,38 @@ public static function allowedSortFields($Allowed = null) {
 
         return self::$AllowedSortFields;
     }
+
+
+    /**
+     * Tests whether a user has permission to view a specific discussion by checking category-specific permissions.
+     * Fires an event that can override the category view permission.
+     *
+     * @param object|integer $discussion The discussion ID or the discussion to test.
+     * @param string|integer $userId The ID of the user to test permission for. If empty, it defaults to Session user.
+     * @param string $categoryPermission The category permission to test against the user.
+     * @return bool Whether the user can view the discussion.
+     * @throws Exception
+     */
+    public function canViewDiscussion($discussion, $userId = '', $categoryPermission = 'Vanilla.Discussions.View') {
+        // Default to session user.
+        if (!$userId) {
+            $userId = val('UserID', Gdn::session(), false);
+        }
+        // Fetch discussion.
+        if (is_numeric($discussion)) {
+            $discussion = $this->getID($discussion);
+        }
+        $userModel = new UserModel();
+        // Get category permission.
+        $canView = $userId && $userModel->getCategoryViewPermission($userId, val('CategoryID', $discussion), $categoryPermission);
```

Comment:
This is wrong. You need to grab the category from CategoryID and then check against its PermissionCategoryID.